Multipart layer fetch

azr opened the pull request:
TLDR: this makes pulls of big images ~2x faster (edit: and a bit more in the latest iteration), and closes #9922. 

cc: #8160, #4989
___ 

Hello Containerd People, I have this draft PR I would like to get your eyes on.

It basically makes pulls faster, but also tries to have not such a big memory impact, by getting consecutive chunks of the layers and immediately pushing them in the pipe (that writes to a file + that signature checksum thing).
I noticed it made pulls ~2x faster, when using the correct settings.

The settings have a big impact, and so I did a bunch of perf tests with different settings, here are some results on a ~8GB image using a `r6id.4xlarge` instance, pulling it from s3.
Gains are somewhat similar on a ~27GB and a ~100GB image (with a little tiny bit of slowdown)
I also tried on an nvme, and a ebs drives, they are ofc slower but gains are still the same.


___

Metrics on a `r6id.4xlarge` timing `crictl pull` of a 8.6GB image.

The first one with 13 tries is with 0 parallelism, it's the current code.
The rest are tries with different settings
* c_para (max number of chunks being pulled per layer at once)
* chunk_size_mb ( size of chunks in mb )
* ctd_max_con ( max # of layer pulled at once )

<details> 
<summary>tmpfs tests:</summary>

```
dst    agv_time          count(*)
-----  ----------------  --------
tmpfs  44.0761538461539  13      

dst    c_para  chunk_size_b  ctd_max_con  agv_time  count(*)
-----  ------  ------------  -----------  --------  --------
tmpfs  110     32            3            22.625    4       
tmpfs  100     32            3            22.64     5       
tmpfs  130     32            2            22.76     1       
tmpfs  120     32            4            22.824    5       
tmpfs  110     32            2            22.85     1       
tmpfs  80      32            4            22.99     1       
tmpfs  110     32            4            23.018    5       
tmpfs  90      64            4            23.09     1       
tmpfs  90      32            3            23.18     1       
tmpfs  110     64            3            23.2125   4       
tmpfs  80      64            3            23.29     1       
tmpfs  90      64            3            23.32     1       
tmpfs  100     32            4            23.352    5       
tmpfs  70      15            4            23.4      1       
tmpfs  100     64            3            23.65     5       
tmpfs  120     15            3            23.68     1       
tmpfs  110     64            2            23.74     1       
tmpfs  100     64            4            23.77     5       
tmpfs  70      32            4            23.81     5       
tmpfs  120     32            3            23.83     5
[...]
```
</details>

<details> 
<summary>nvme (885GB) tests:</summary>

```
dst         agv_time          count(*)
----------  ----------------  --------
added-nvme  47.4008333333333  12      

dst         c_para  chunk_size_mb  ctd_max_con  agv_time  count(*)
----------  ------  ------------  -----------  --------  --------
added-nvme  130     32            3            25.24     1       
added-nvme  70      32            4            26.1      1       
added-nvme  80      32            3            26.31     1       
added-nvme  100     32            3            26.38     1       
added-nvme  120     32            4            26.58     1       
added-nvme  130     32            2            26.71     1       
added-nvme  80      32            4            26.73     1       
added-nvme  120     10            3            26.82     1       
added-nvme  80      64            3            26.93     1       
```
</details>

____

Observations, I did a little go program to multipart download big files directly into a file at different positions with different requests, and that was much faster than piping single threadedly into a file. Containerd pipes in a checksumer and then pipes into a file. I think that this can in some conditions create some sort of thrashing, hence why the parameters are very important here.

That simple go program had pretty bad perfs with one connection, but I was able to saturate the network with multiple connection, with better or on par perfs with [aws-crt](https://github.com/awslabs/aws-c-s3).

I think that for maximum perfs, we could try to re-architecture things a bit; like concurrently write directly into the tmpfile, and then tell the checksumer our progress, so that it can do that in parallel, and then carry on like usual.

swagatbora90 commented:
@azr Thanks for the PR, this looks promising. I wonder if you were able to get any memory usage data from your tests? Previous effort to use [ECR containerd resolver ](https://github.com/awslabs/amazon-ecr-containerd-resolver) showed that it can take up disproportionate amount of memory specially when we increase the number of parallel chunks(without providing significant benefit to latency). The high memory utilization was mainly from the `htcat` library that ECR resolver uses to performance parallel Ranged Gets. I think we should understand these tradeoffs. 

Also can you share some information about your test image? Number of layers? Size of individual layers?

akhilerm commented:
/ok-to-test

azr commented:
Hey, sorry for the double post, I have a config question, there are 3 places I could think of, to put the config in for this and I keep second guessing myself as to where would be the most appropriate place, I'd like to pick the best possible decision here.

I can think of:
1/ `[plugins."io.containerd.transfer.v1.local"]`

This seems doable, and the `max_concurrent_downloads` setting is in there, but there is this TODO comment:

https://github.com/containerd/containerd/blob/7b5df7e41a95520cbdbe02e9d4c5d625f618ae1d/internal/cri/config/config.go#L294-L296

2/ `[plugins."io.containerd.grpc.v1.cri"]` setting

Also has a `max_concurrent_downloads` setting, but that doesn't seem to be it, and the local service says to use the transfer service. Also, the fetcher has no reference to this configuration. Is there an idiomatic way to get that setting without messing the cfg architecture up ?

3/ lastly some per-registry setting eg: `[plugins."io.containerd.grpc.v1.cri".registry.configs."my.registry.io".auth]`

This would be easy to implement and great to allow to have a per-registry setting, but it still would be nice to have and option 1,2 to default to.
___
My favorite option is to go with 2 + 3, but, WDYT ?

swagatbora90 commented:
@azr Thanks for adding the performance numbers. I ran some tests as well using your patch and the memory usage looks better than what I saw in the htcat implementation specially with high parallelism count.

However, I do observe that increasing parallelism does not yield better latency and may lead to higher memory usage (I think there is a number of other factors to consider here mainly type of instance used for testing, network bandwidth). I tried to limit the test to a single image with a single layer and fixing the chunk size to 20 MB. A lower parallelism count(3 or 4) may be preferable than setting parallelism to upwards of 10. 

Using a c7.12xlarge instance to pull a 3GB single layer image from ECR private repo.

| Parallelism  Count| Chunk Size(MB) | Total Download time(sec) | Network Pull time(sec) | Download Speed(MBPS) | Max Memory used (from cgroups memory.peak)  |
|------------|-------------|---------------------|-------------------|------------------|--------------------| 
|     1             |      20           |            65.9               |           51.78            |       53                 |       15.9|
|2|20|39.3|32.08|88.8|17.5|
|3|20|36.6|22.57|95.4|18|
|4|20|36.8|16.82|94.8|17|
|5|20|36.8|14.78|94.8|17|
|10|20|36.9|13.92|94.6|20|
|20|20|36.9|14.31|94.6|22|
|30|20|36.7|14.97|95.1|26|
|40|20|36.8|14.21|94.8|31|
|50|20|36.7|14.3|95.1|36|
|100|20|36.8|14.91|94.8|52

![multipart1](https://github.com/containerd/containerd/assets/5525370/dbd5d6da-7b36-45cd-85f1-35dd6400fd71)


Also the network download time was much faster (see Network Pull time) (~15sec) while containerd took additional ~20secs to complete the pull (before it started unpacking). I calculated the Network Download time by periodically calling `/containerd.services.content.v1.Content/ListStatuses"` filtering the layer digest and checking when the content.Offset == content.Size. I am still not sure why containerd takes so much time after it has already committed to the content store, pprof does not show any significant cpu usage by containerd either during this time. Are we blocked on GC or some underlying syscall(fp.Sync) to complete?

swagatbora90 commented:
@dmcgowan @kzys

dmcgowan commented:
Thanks @azr, the numbers on this look good. @swagatbora90 super helpful stats.

We should continue to respect `max_concurrent_downloads` as the upper limit for actively downloading connections. Can we make the parallelism dynamic, maybe use [TryAcquire](https://pkg.go.dev/golang.org/x/sync/semaphore#Weighted.TryAcquire) until we hit the limit to determine the parallelism. It definitely makes sense to use the concurrency allowance to download the first layers faster rather than spreading across all layers, since unpack still takes up a significant amount of the pull time.

For configuration, let's use the transfer service configuration. This won't make it in for 2.0 and CRI will be switching to transfer service by 2.1.

azr commented:
Hey @dmcgowan ! Nice, thanks.
Good idea to keep `max_concurrent_downloads` docs/behaviour consistent. I was poking around your suggestion to lock N goroutines, and the thing is `max_concurrent_downloads` also controls how many unpacks are happening at once because these are all being called in stacked handlers. 

I have options in mind, and have to think/test of a good way to do this. I might have to introduce a `max_concurrent_unpacks` setting. On it ! 😊

azr commented:
Hello there ! Sorry for the radio silence, had some emergencies.
I come with good news ! I have a new build that is even faster.
I now can see downloads of `~15s` for my `8.6GB` image, not always. ( original pull time is `~45s`, first iteration was `~25s` )

This modifies the local transfer plugin to allow for the following settings:

```
  # setting that is great for me:
  [plugins."io.containerd.transfer.v1.local"]
   max_concurrent_downloads = 444
   max_concurrent_download_operations = 3 
   max_concurrent_downloads_per_layer = 111
   concurrent_fetch_chunks_size_mb = 30
```

Previously, the `max_concurrent_downloads` was also limiting the number of operation that we could run (dl, unpack and mv) as the weighted semaphore was used in the handler loop. I added a new `max_concurrent_download_operations` setting that does that instead, with a new semaphore.
The `max_concurrent_downloads` semaphore is passed (pointer) to the fetch, so that it can do its thing at pull, and only is used during actual downloads.

Note that during my tests I used a `r6id_4xlarge` machine with a vpc directly peered to s3. The backend for this is distribution, and distribution returns s3 dl link. s3 has great perfs. Pulling from *the* docker registry is way slower, probably due to some rate limiting.

azr commented:
Is this test a flaky one? (cc: @swagatbora90 @dmcgowan ) Looks like this is the last one. 

LMK what you think about this PR ! ( Especially the part where I change some defaults )

```
=== FAIL: plugins/snapshots/windows TestWindows/MoveFileFromLowerLayer (0.22s)
    log_hook.go:47: time="2024-06-26T08:42:33.715525600Z" level=debug msg=createSnapshot func="windows.(*wcowSnapshotter).createSnapshot.func1" file="D:/a/containerd/containerd/src/github.com/containerd/containerd/plugins/snapshots/windows/windows.go:152" testcase=TestWindows/MoveFileFromLowerLayer
    testsuite.go:852: Check snapshots failed: failed to create snapshot 1: failed to apply: failed to mount: failed to set volume mount path for layer C:\Users\RUNNER~1\AppData\Local\Temp\snapshot-suite-Windows-3182491901\root\snapshots\1: failed to bind target "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\snapshot-suite-Windows-3182491901\\root\\snapshots\\1\\Files" to root "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\snapshot-suite-Windows-3182491901\\work\\prepare1984866504": Incorrect function.
    helpers.go:71: drwxrwxrwx          0 C:\Users\RUNNER~1\AppData\Local\Temp\snapshot-suite-Windows-3182491901
    helpers.go:71: drwxrwxrwx          0 C:\Users\RUNNER~1\AppData\Local\Temp\snapshot-suite-Windows-3182491901\root
    helpers.go:69: -rw-rw-rw-      65536 C:\Users\RUNNER~1\AppData\Local\Temp\snapshot-suite-Windows-3182491901\root\metadata.db [ "\x00\x00\x00\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00\x00\xed\xda\f\xed\x02\x00\x00\x00\x00\x10\x00\x00\x00\x00\x00\x00\a\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\b\x00\x00\x00\x00\x00\x00\x00\t\x00\x00\x00\x00\x00\x00\x00" ...]
    helpers.go:71: drwxrwxrwx          0 C:\Users\RUNNER~1\AppData\Local\Temp\snapshot-suite-Windows-3182491901\root\snapshots
    helpers.go:71: drwxrwxrwx          0 C:\Users\RUNNER~1\AppData\Local\Temp\snapshot-suite-Windows-3182491901\root\snapshots\1
    helpers.go:71: drwxrwxrwx          0 C:\Users\RUNNER~1\AppData\Local\Temp\snapshot-suite-Windows-3182491901\root\snapshots\1\Files
    helpers.go:71: drwxrwxrwx          0 C:\Users\RUNNER~1\AppData\Local\Temp\snapshot-suite-Windows-3182491901\work

=== FAIL: plugins/snapshots/windows TestWindows (0.01s)

DONE 1034 tests, 10 skipped, 2 failures in 560.641s
mingw32-make: *** [Makefile:210: test] Error 1
Error: Process completed with exit code 2.
```

azr commented:
/retest

swagatbora90 commented on client/client.go:
```diff
@@ -369,6 +369,18 @@ type RemoteContext struct {
 	// MaxConcurrentDownloads is the max concurrent content downloads for each pull.
 	MaxConcurrentDownloads int
 
+	// MaxConcurrentDownloadOperations is the max number of operation that can
```
These are defined but not used. Do you plan to add support for client opts as a follow up?

azr replied:
not at all, an artefact from when I wasn't sure if these params needed to go there. Removing. I can add support for the client if you think it's worth it though

swagatbora90 replied:
I think this will be useful to have specially for clients that do not use transfer service for image related operations. Though this can be done as a follow up PR and need not be part of this PR.

azr replied:
Gotcha, awesome, yes, I can do a follow up 😊

azr replied:
Okay, will spend some time to do this one.

azr replied:
Hello, I re-added client support, please lmk what you think !

With this addition, I've been able to pull with these settings through `crictl`

swagatbora90 commented on core/remotes/docker/fetcher.go:
```diff
@@ -262,7 +265,31 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest, op
+)
+
 func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string, offset int64) (_ io.ReadCloser, retErr error) {
+	parallelism := int64(1)
```
Nit: if parallelism = 1 is the default behavior, it may be useful to define this value as constants  in the beginninga long with a coment.

azr replied:
Good idea, adding this.

swagatbora90 commented on core/remotes/docker/fetcher.go:
```diff
@@ -262,7 +265,31 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest, op
+
 func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string, offset int64) (_ io.ReadCloser, retErr error) {
+	parallelism := int64(1)
+	if r.config.MaxConcurrentDownloads > 0 {
```
This will change the behavior of existing clients which do not set the `MaxConcurrentDownloadsPerLayer` but only set the `MaxConcurrentDownloads` value in their client options. I think if a client does not set `MaxConcurrentDownloadsPerLayer` value, the behavior should default to current approach of single fetch per layer. Thoughts?

azr replied:
Agreed ! Better to remain consistent 😊, updating.

azr replied:
Hm, totally I agree here ! Changing this, so that this feature is opt-in !

swagatbora90 commented on core/remotes/docker/fetcher.go:
```diff
@@ -318,19 +352,110 @@ func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string,
 		return r == ' ' || r == '\t' || r == ','
 	})
+
+	minimumSizeForParallelDL := 1 * mB
```
Similar. Prefer these values to be defined as constants

swagatbora90 commented on plugins/transfer/plugin.go:
```diff
@@ -176,7 +194,10 @@ type unpackConfiguration struct {
 	return &transferConfig{
-		MaxConcurrentDownloads:      3,
-		MaxConcurrentUploadedLayers: 3,
+		MaxConcurrentDownloads:          9,
```
Wondering why we decided to set this default in transfer service? The local (non-transfer service) client implementation has a different behavior where all layers are pulled in parallel if `WithMaxConcurrentDownloads` is not set.

azr replied:
Reverting this, I wanted to change the default behaviour of containerd to always use parallel layer fetching with a safe default, but maybe this can be done in a subsequent PR.

swagatbora90 commented on core/transfer/transfer.go:
```diff
@@ -127,6 +128,48 @@ func WithProgress(f ProgressFunc) Opt {
 	}
 }
 
+type FetcherConfig struct {
```
isn't `remotes/resolver.go` a better place to define these?

azr replied:
This was not a possibility as `core/remotes` imports `core/transfer` and I needed to pass these options to `transfer.ImageFetcher`. Creating an import cycle.

azr replied:
https://github.com/azr/containerd/blob/dcd4b1234d11c43066fc7e81718744819b5a52f8/core/transfer/transfer.go#L38-L42

azr replied:
After thinking a bit, I'm feeling like this is okay like this, but lmk !

swagatbora90 replied:
Feel free to ignore my suggestion here.

My thinking was we define the new FetcherOpt directly in remotes/resolver.go instead of in transfer.
So the Resolver interface changes to
```
type Resolver interface {

-       Fetcher(ctx context.Context, ref string, opts ...transfer.FetcherOpt) (Fetcher, error)
+       Fetcher(ctx context.Context, ref string, opts ...FetcherOpt) (Fetcher, error)
```
So, remotes does not import transfer (importing fetcherOpts from transfer felt a bit weird)
And in the transfer package, the ImageFetcher interface can be changed to:
```
 type ImageFetcher interface {
        ImageResolver
 
-       Fetcher(ctx context.Context, ref string, opts ...FetcherOpt) (Fetcher, error)
+       Fetcher(ctx context.Context, ref string, opts ...remotes.FetcherOpt) (Fetcher, error)
```
There should not be any cyclic dependencies since package resolver or package docker will not have a dependency on transfer, but transfer will depend on package transfer.

azr replied:
Ah, very good point, this is way better, thanks !

swagatbora90 commented on core/transfer/local/transfer.go:
```diff
@@ -54,8 +56,17 @@ func NewTransferService(cs content.Store, is images.Store, tc TransferConfig) tr
 		ts.limiterU = semaphore.NewWeighted(int64(tc.MaxConcurrentUploadedLayers))
 	}
+	if tc.MaxConcurrentDownloadsPerLayer > 0 {
+		ts.limiterU = semaphore.NewWeighted(int64(tc.MaxConcurrentDownloadsPerLayer))
```
Is this correct? why is `limiterU` value being reset here?

azr replied:
good catch, this is not correct; fixing.

azr replied:
typo copy paste, good catch, this is quite wrong, removing

vrdn-23 commented:
Thanks for the awesome work @azr! Just wanted to know if this was still on your radar?

azr commented:
Hello ! Yes it is ! Was on a big vacation, will start up on monday.

azr commented on core/remotes/docker/fetcher.go:
```diff
@@ -39,6 +41,7 @@ import (
 
 type dockerFetcher struct {
 	*dockerBase
+	config transfer.FetcherConfig
```
see https://github.com/containerd/containerd/pull/10177#discussion_r1668137145, tldr it was to avoid an import cycle issue

azr commented on plugins/transfer/plugin.go:
```diff
@@ -153,6 +157,20 @@ type transferConfig struct {
+	// MaxConcurrentDownloadsPerLayer is the max concurent download per layer
+	// for a pull. 1 means a layer will be donloaded with one connection. 0
+	// means no limit, the upper limit will then be `max_concurrent_downloads`.
+	MaxConcurrentDownloadsPerLayer int `toml:"max_concurrent_downloads_per_layer"`
```
Added some words + about default ( saying this is off by default )

azr commented:
Hello @swagatbora90 thanks a lot for your thoughtful review 🙏🏼🙏🏼🙏🏼🙏🏼, I think I addressed most things, please let me know if there is anything I can change/fix/improve.

azr commented:
/retest

azr commented:
/retest

azr commented:
/retest-required

kzys commented on core/remotes/docker/fetcher.go:
```diff
@@ -262,7 +265,27 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest, op
+		if err := r.config.Limiter.Acquire(ctx, parallelism); err != nil {
+			return nil, err
+		}
+	}
```
Nit: How about having a defer block to do `r.config.Limiter.Release(parallelism)` here?

kzys reviewed (commented):
Looks good to me! One nit

azr commented:
/retest

azr commented:
/retest

azr commented:
Little update here, I have been using `syne-tune` to get great parameters for my setup, and got the download time of that 27Gb image to a tmpfs down to `6.968424s`, probably a few lucky runs, but still nice numbers !!

cartermckinnon commented on core/remotes/docker/fetcher.go:
```diff
@@ -274,11 +297,14 @@ func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string,
 		// will return the header without supporting the range. The content
 		// range must always be checked.
-		req.header.Set("Range", fmt.Sprintf("bytes=%d-", offset))
+		req.header.Set("range", fmt.Sprintf("bytes=%d-", offset))
```
Why'd you change the case of the HTTP headers (in a few places)?

azr replied:
Will make everything 'canonical' to avoid operations and for consistency.

https://github.com/golang/go/blob/d36353499f673c89a267a489beb80133a14a75f9/src/net/textproto/reader.go#L647-L666

vrdn-23 commented:
@kzys @swagatbora90 @cartermckinnon 
Just wanted to check-in and find out if this PR was good to get merged in? The improvements highlighted by @azr are something definitely to look forward to!

azr commented:
Hello there, another update.

I updated the cri (client) part to be able to use the same settings. It works, is faster, but is somehow slower than using the local service. Have to double check everything.

Moreover, I triple checked the timings of my previous post ( `44s` to `7s` ), and it's very rare to get such a speedup. But, in general and using averages, we remain in the x2 window. 
Using files instead of streaming would probably improve things a bit further here.

BTW, I read [here](https://github.com/containerd/containerd/issues/7592) that the CRI plugin should soon start using the local transfer plugin ? Is there a PR somewhere with that ? Would it be something I could do as well ?

azr commented:
/retest

azr commented:
Ok, I think found out why `ctr pull` was slower than `crictl pull`. there seems to be a bug in `ctr pull`, that I fixed in my last commit. No unpacking was performed by default, unless an architecture was selected. But an arch was selected. I just had to change the order of execution, for the unpacking to run.

This PR should be good to review and to merge (to me), sorry for the wait.

cc: @kzys, @swagatbora90, @cartermckinnon

Context: since you last reviewed I added client setting for multipart pull. LMK if you'd like me to change anything.

vrdn-23 commented:
@kzys , @swagatbora90 , @cartermckinnon just wanted to check in and know if this PR is going to be prioritized by the team  or is there an alternate approach that is being considered?

djdongjin commented on client/client.go:
```diff
@@ -395,6 +395,15 @@ type RemoteContext struct {
 	// MaxConcurrentDownloads is the max concurrent content downloads for each pull.
 	MaxConcurrentDownloads int
 
+	// MaxConcurrentDownloadOperations is the max concurrent download operations for each pull.
+	MaxConcurrentDownloadOperations int
+
+	// MaxConcurrentDownloadsPerLayer is the max concurrent downloads per layer for each pull.
+	MaxConcurrentDownloadsPerLayer int
```
Can we add some explanation about the difference among the 3 options, especially the first two options? thanks.

E.g., my understading is that before this PR, an image pull will simply have `MaxConcurrentDownloads` number of concurrent layer download (one for each layer). What will be the behavior after thie PR?

SaranBalaji90 replied:
Why do we need MaxConcurrentDownloadOperations?

azr replied:
I updated the docs ! LMK if that helps better !

djdongjin commented on plugins/transfer/plugin.go:
```diff
@@ -153,6 +157,21 @@ type transferConfig struct {
 	// MaxConcurrentDownloads is the max concurrent content downloads for pull.
 	MaxConcurrentDownloads int `toml:"max_concurrent_downloads"`
 
+	// MaxConcurrentDownloadsPerLayer is the max concurent download per layer
```
`max concurent` -> `max concurrent`

djdongjin commented on plugins/transfer/plugin.go:
```diff
@@ -153,6 +157,21 @@ type transferConfig struct {
 	MaxConcurrentDownloads int `toml:"max_concurrent_downloads"`
 
+	// MaxConcurrentDownloadsPerLayer is the max concurent download per layer
+	// for a pull. 1 means a layer will be donloaded with one connection. 0
```
`be donloaded` -> `be downloaded`

djdongjin commented on core/remotes/resolver.go:
```diff
@@ -109,3 +113,62 @@ func WithMediaType(mediatype string) FetchByDigestOpts {
+}
+
+func (cfg FetcherConfig) Parallelism() int64 {
+	// parallel layer fetching remains off when MaxConcurrentDownloadsPerLayer
+	// is set to 0. This could be changed in the future if/when the feature is
+	// successful and widely adopted.
```
I think we should change this to `parallel layer fetching remains off if MaxConcurrentDownloadsPerLayer <= 1` because:

1. `MaxConcurrentDownloadsPerLayer=1` equals to no parallel layer fetching.
2. Currently 1 can be returned in 2 places: line 133 if `MaxConcurrentDownloadsPerLayer<= 0` and line 136 if `MaxConcurrentDownloadsPerLayer=1`, which is confusing.

WDYT?

azr replied:
That makes more sense, yes, updating.

SaranBalaji90 commented on core/remotes/docker/fetcher.go:
```diff
@@ -318,19 +347,108 @@ func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string,
+						continue
+					}
+					reqClone := req.clone()
+					reqClone.header.Set("Range", fmt.Sprintf("bytes=%d-", offset+i*chunkSize))
```
Thank you for the PR. I was glancing through it out of curiosity (may have more questions later). Why would we set just the start of the range but not the end?

azr replied:
Hm, I just never did it, mainly because it felt like we could close the connection when done and avoid having to calculate that. Do you think this would be important ?

SaranBalaji90 commented on core/remotes/docker/fetcher.go:
```diff
@@ -318,19 +347,108 @@ func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string,
+		stopChan := make(chan struct{})
+		readers, writers := make([]io.Reader, numChunks), make([]*io.PipeWriter, numChunks)
+		for i := int64(0); i < numChunks; i++ {
+			readers[i], writers[i] = io.Pipe()
```
Did you happen to evaluate other mechanism to download chunks? Since this uses io.pipe, it was somewhat slower on my tests since writers were all paused until readers read those chunks in order. I was able to improve the performance significantly (will post numbers shortly) by storing all these chunks in a temp directory and reading them after all downloads are complete. 

```
                tempDir, _ := os.MkdirTemp("/tmp", "digest")
		var wg sync.WaitGroup
		for p := int64(0); p < parallelism; p++ {
			wg.Add(1)
			// start parallel download workers
			...
					if i == 0 {
						_, err = io.Copy(f, io.LimitReader(resp.Body, chunkSize))
						_ = resp.Body.Close()
						continue
					}
					...
					_, err = io.Copy(f, nresp.Body)
					_ = nresp.Body.Close()
					...
			}()
		}
		wg.Wait()
		body = getSortedPartsReader(tempDir)
		...
		
func getSortedPartsReader(tempDir string) io.ReadCloser {
	matches, _ := filepath.Glob(filepath.Join(tempDir, "*.part*"))
	c := collate.New(language.English, collate.Numeric)
	c.SortStrings(matches)
	readers := make([]io.Reader, len(matches))
	for i, filename := range matches {
		file, _ := os.Open(filename)
		readers[i] = file
	}
	return io.NopCloser(io.MultiReader(readers...))
}
```

azr replied:
I also had a much faster test that was downloading in a file with punched holes, that was used as a buffer.
I like your idea to use parts like this too, but this feels like adding a new set of feature to download into a tmp dir.

Would that be okay to do that?
Should we cleanup the folder afterwards ?

azr replied:
We could add a part to auto remove files onces downloaded. We will have to be careful with files that don't exist yet as `getSortedPartsReader` will list what is currently there, but not the rest.

azr replied:
I personally feel like we should merge this PR, and then I will try to do the punched-buffer-file thing.

azr replied:
Hello, turns out we worked on this with @co42, we are using buffered pipes ( io doesn't have these somehow ) and we are getting close to speeds when using a buffer.

SaranBalaji90 commented on core/remotes/docker/fetcher.go:
```diff
@@ -318,19 +347,108 @@ func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string,
+
+	totalSize, _ := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
+	remaining := totalSize - offset
+	if parallelism > 1 && chunkSize > 0 && remaining > minimumSizeForParallelDL && req.body == nil {
```
Just FYI, here is similar piece of software to help clean up these conditions a bit - https://github.com/aws/aws-sdk-go-v2/blob/main/feature/s3/manager/download.go#L245

azr replied:
Do you mean for inspiration ?

SaranBalaji90 commented on core/remotes/docker/fetcher.go:
```diff
@@ -318,19 +347,108 @@ func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string,
+					default:
+					}
+					if i == 0 {
+						_, err = io.Copy(writers[i], io.LimitReader(resp.Body, chunkSize))
```
What about the error handling here?

azr replied:
The writer is closed with the error a bit later:

```
writers[i].CloseWithError(err)
```

In the unit tests I added it was passed along.

SaranBalaji90 commented on core/remotes/docker/fetcher.go:
```diff
@@ -262,7 +264,27 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest, op
+	minimumSizeForParallelDL = 1 * mB
+)
+
 func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string, offset int64) (_ io.ReadCloser, retErr error) {
```
Can you help modularize this function, its already somewhat big enough to break it down.

azr replied:
yes !

azr replied:
Hm, I tried a bit, and in this case I think breaking it down would only complicate things and make it harder to understand, etc.

Do you confirm you'd really like me to split it up ?

kzys replied:
I'm fine without splitting the function.

Is it possible to do `r.config.Limiter.Release(parallelism)` in defer? Would it make this function simpler? Not a blocker, but I'm curious.

azr replied:
it would not be possible in a defer here, because we acquire the log 'greedily', like we ask `parallelism`, but if we cannot get `parallelism` we release the diff.
Also we defer release 1 as soon as a goroutine is done.

Would you like this behaviour to be different ?

cartermckinnon replied:
I can see why this is the way it is, but touching the semaphore in so many places does seem like bug fodder. Could we add a condition within the `defer` or something?

azr replied:
Gotcha, trying to have something better with semaphores while taking into account your other comment. 🤔🤔🤔🤔

azr replied:
Okay, I pushed something different with better error handling, more granular semaphore calls and defers, LMK

sadathknorket commented:
Hi , will this be merged ?

kzys commented on client/client.go:
```diff
@@ -392,9 +392,30 @@ type RemoteContext struct {
 	MaxConcurrentDownloads int
 
+	// MaxConcurrentDownloadOperations limits how many operations can run in
+	// parallel during an image pull. Operations include: - Downloading a layer
```
nit: seems Markdown list items are squashed by gofmt or something.

azr replied:
fixed, thx

kzys commented on cmd/ctr/commands/images/pull.go:
```diff
@@ -120,16 +120,17 @@ command. As part of this process, we do the following:
 				sopts = append(sopts, image.WithPlatforms(p...))
 			}
+
+			// Set unpack configuration
```
Why did you move this section?

azr replied:
(reverted to main now )

My fix was the following: https://github.com/containerd/containerd/commit/79a42eedc724cd248a995cbf1174d3800d948d52

I wasn't understanding why my pull was so fast, and why in kube so slow, turns out a pull was not unpacking by default.

Someone else merged a fix before me, so, reverted.

kzys commented on core/remotes/docker/resolver.go:
```diff
@@ -554,6 +555,13 @@ type request struct {
+func (r *request) clone() *request {
+	res := *r
+	res.header = r.header.Clone()
+	//res.host.Header = r.host.Header.Clone()
```
Don't need this commented-out line?

azr replied:
good point, removed !

azr commented:
/retest

azr commented:
/retest

cartermckinnon commented:
Failures do not look related to this PR:
```
   [FAIL] [sig-node] Kubelet Authz [Feature:KubeletFineGrainedAuthz] when calling kubelet API [It] check /healthz enpoint is accessible via nodes/healthz RBAC [sig-node, Feature:KubeletFineGrainedAuthz]
  k8s.io/kubernetes/test/e2e_node/kubelet_authz_test.go:82
  [FAIL] [sig-node] Kubelet Authz [Feature:KubeletFineGrainedAuthz] when calling kubelet API [It] check /healthz enpoint is not accessible via nodes/configz RBAC [sig-node, Feature:KubeletFineGrainedAuthz]
  k8s.io/kubernetes/test/e2e_node/kubelet_authz_test.go:82
  [FAIL] [sig-node] Kubelet Authz [Feature:KubeletFineGrainedAuthz] when calling kubelet API [It] check /healthz enpoint is accessible via nodes/proxy RBAC [sig-node, Feature:KubeletFineGrainedAuthz] 
```

/retest

cartermckinnon commented on core/remotes/docker/fetcher.go:
```diff
@@ -262,7 +264,27 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest, op
 func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string, offset int64) (_ io.ReadCloser, retErr error) {
+	parallelism, chunkSize := r.config.Parallelism(), int64(r.config.ConcurrentFetchChunksSizeMB)*mB
+	if r.config.Limiter != nil {
+		if err := r.config.Limiter.Acquire(ctx, parallelism); err != nil {
```
So we won't start doing any work until we can fire on all cylinders? Would it make sense to use `TryAcquire` and accept a lower parallelism if we can avoid waiting?

azr replied:
That would work but we would have a chance of downloading the biggest layer of the whole set without going full throttle. Which could randomly lower speed by a lot. Trying to see if I can have a setup were we can have both.

cartermckinnon reviewed (commented):
Thanks for the progress @azr! I think this is getting close.

azr commented:
Update here, we changed a bit of things and the PR should look great now. Semaphore granularity + defer fixes, @co42 added a buffered pipe, that helps a lot.


/retest

azr commented:
/retest

kzys commented on client/client.go:
```diff
@@ -394,9 +394,34 @@ type RemoteContext struct {
+	// when downloading layers in parallel (requires
+	// MaxConcurrentDownloadsPerLayer > 1). Larger chunks reduce coordination
+	// overhead but use more memory.
+	ConcurrentFetchChunksSizeMB int
```
How about passing the size in bytes, instead of MB? We have some `Size` variables and most (all?) of them are in bytes.

https://github.com/search?q=repo%3Acontainerd%2Fcontainerd+size&type=code

azr replied:
Good point, updated, after the buffered pipe we need to be more subtle with this setting

azr commented on core/remotes/docker/fetcher.go:
```diff
@@ -35,10 +37,168 @@ import (
 
+type BufferPool struct {
+	sync.Mutex
+	sem    semaphore.Weighted
```
updated

azr replied:
good point

azr commented:
Updated @kzys thanks for your reviews !! LMK if anything.

azr commented:
not 100% certain failure is related:

```
Summarizing 1 Failure:
  [FAIL] [sig-node] ResourceMetricsAPI [Feature:ResourceMetrics] when querying /resource/metrics [It] should report resource usage through the resource metrics api [sig-node, Feature:ResourceMetrics]
  k8s.io/kubernetes/test/e2e_node/resource_metrics_test.go:132
Ran 375 of 734 Specs in 1246.139 seconds
FAIL! -- 374 Passed | 1 Failed | 0 Pending | 359 Skipped
```

will try to double-check anyways

/retest

azr commented:
different failures, let's see

/retest

djdongjin commented on client/client.go:
```diff
@@ -394,9 +394,34 @@ type RemoteContext struct {
+	// when downloading layers in parallel (requires
+	// MaxConcurrentDownloadsPerLayer > 1). Larger chunks reduce coordination
+	// overhead but use more memory.
+	ConcurrentFetchChunksSize int
```
nit: suggest to call it `ConcurrentDownloadChunkSize` (`Fetch` -> `Download` to be consistent with other related config) (`Chunks` -> `Chunk` as it specifies size of a single chunk)

azr replied:
Good point ! Updating ( and the rest too )

djdongjin commented on core/transfer/local/transfer.go:
```diff
@@ -54,8 +56,14 @@ func NewTransferService(cs content.Store, is images.Store, tc TransferConfig) tr
+	}
 	if tc.MaxConcurrentDownloads > 0 {
 		ts.limiterD = semaphore.NewWeighted(int64(tc.MaxConcurrentDownloads))
+	} else {
+		// default to 3 dl at a time
+		ts.limiterD = semaphore.NewWeighted(int64(3))
```
why do we need a default 3 semaphore now? can we still keep the previous behavior if tc.MaxConcurrentDownloads == 0?

azr replied:
The default was 3, so, using this as the default, ex:

```
--max-concurrent-downloads int            Set the max concurrent downloads for each pull (default 3)
```

I changed this behaviour because, before this PR `max-concurrent-downloads` was creating a semaphore that was limiting the number of operations (copy/unpack/else) AND the number of downloads. With this PR, we are still limiting to 3 downloads at a time, but the operation semaphore is another one.

azr replied:
That operations semaphore is not limited by default `MaxConcurrentDownloadOperations`, and maybe I should put it to 1 to keep the old behaviour ? I think that 0 is also fine here, for a setting with 3 downloads at a time.

azr replied:
After a bunch of re-reading, I can put back the default. But the user will have to not set `MaxConcurrentDownloadsPerLayer`, otherwise we might use way too much ram

djdongjin commented on core/remotes/docker/fetcher.go:
```diff
@@ -35,10 +37,168 @@ import (
+	"golang.org/x/sync/semaphore"
 )
 
+type BufferPool struct {
+	sync.Mutex
+	sem    *semaphore.Weighted
+	pool   []bytes.Buffer
+	bufCap int64
+}
```
can we use `sync.Pool`(https://pkg.go.dev/sync#Pool) directly? There are some examples of this in containerd code.

azr replied:
good point, updated !

djdongjin commented on core/remotes/docker/fetcher.go:
```diff
@@ -35,10 +37,168 @@ import (
+
+var ErrClosedPipe = errors.New("bufpipe: read/write on closed pipe")
+
+type pipe struct {
+	cond       *sync.Cond
+	bufPool    *BufferPool
+	buf        *bytes.Buffer
+	rerr, werr error
+}
```
qq: can we add some brief comments describing what `pipe` is used for? Will be helpful for others to understand below code :) thanks

azr replied:
added a bunch, LMK !

djdongjin commented on core/remotes/docker/fetcher.go:
```diff
@@ -35,10 +37,168 @@ import (
+	rerr, werr error
+}
+
+type PipeReader struct {
+	*pipe
+}
+
+type PipeWriter struct {
+	*pipe
+}
```
these can be unexported?

azr replied:
yes!

djdongjin commented on core/remotes/resolver.go:
```diff
@@ -109,3 +113,77 @@ func WithMediaType(mediatype string) FetchByDigestOpts {
+// Layer fetch config
+////
+
+type FetcherConfig struct {
```
does this need to be exported?

azr replied:
Yes, it's passed in `core/remotes/docker.` by `core/remotes` to pass the download settings/limitations.

azr commented:
Updated @djdongjin, thanks for reviewing, lmk if anything.

/retest

azr commented:
```
  Latency metrics for node tmp-node-e2e-654a8a52-cos-beta-117-18613-164-13
  STEP: Destroying namespace "resource-metrics-2200" for this suite. @ 02/10/25 14:21:18.198
  << Timeline
  [FAILED] Timed out after 60.001s.
```

/retest

djdongjin commented on core/remotes/resolver.go:
```diff
@@ -43,7 +47,7 @@ type Resolver interface {
 	// Fetcher returns a new fetcher for the provided reference.
 	// All content fetched from the returned fetcher will be
 	// from the namespace referred to by ref.
-	Fetcher(ctx context.Context, ref string) (Fetcher, error)
+	Fetcher(ctx context.Context, ref string, opts ...FetcherOpt) (Fetcher, error)
```
This will be a breaking change to this interface.

azr replied:
Yes, hopefully not too bad given the var-args, but yes. But yes, I see what you mean, the "if you say yes, it's forever" kid of breaking change.

djdongjin commented on core/remotes/resolver.go:
```diff
@@ -109,3 +113,77 @@ func WithMediaType(mediatype string) FetchByDigestOpts {
+////
+
+type FetcherConfig struct {
+	MaxConcurrentDownloads         int
+	MaxConcurrentDownloadsPerLayer int
+	ConcurrentDownloadChunkSize    int
+	Limiter                        limiter
```
why does the `FetcherConfig` need to have both `int` limit configs as well as an actual `limiter`?

azr replied:
The limiter is here so that two downloads running at the same time use the same limiter.

The rest is the config, ex, to tell for example how many goroutines could be started at once.

azr replied:
I kinda wanted to wrap the limiter with something that could give it's own limit, to avoid that, and to take the number from it. Didn't do it to not add too much code.

djdongjin commented on core/remotes/docker/fetcher.go:
```diff
@@ -263,74 +439,120 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest, op
-		// range must always be checked.
-		req.header.Set("Range", fmt.Sprintf("bytes=%d-", offset))
-	}
+	req.setMediaType(mediatype)
+	req.header.Set("Accept-Encoding", "zstd;q=1.0, gzip;q=0.8, deflate;q=0.5")
+	req.setOffset(offset)
```
still need some time to digest the changes in this function.

I wonder if we can incapsulate changes into a separate function (that has multi part fetch support), and based on if we need multi-part fetch, we all differeent functions.

azr replied:
Understood, and as you prefer ! 😊 My thinking here is that when parallel layer fetch is off, this is the exact same function, so this should be okay. LMK if anything

azr replied:
Hey @djdongjin did you have enough time to digest these changes ? Are there any things I could do to make merging easier ?

Originally I was looking for sort of a plugin system that would let me create my own 'puller', it would have been nice to be able to switch back and forth.

djdongjin replied:
Hi @azr I will take a look today or early next week.

> Originally I was looking for sort of a plugin system that would let me create my own 'puller'

We actually have something similar :) instead of a plugin it's a separate systemd service. So the client will call the separate `puller` first which will download and put the blobs under containerd content store. So when the client later calls containerd to pull the image it's already there.

But I think it's great to add this to containerd itself and make the parallelism configurable.

azr replied:
Hey @djdongjin ! Coming back to this thread as it's been a bit more than a week. Do you have any new? Thanks !

djdongjin replied:
sorry was quite busy these 2 weeks. I will try to find sometime this week!

djdongjin reviewed (commented):
Left some comments. Still need some time to understand more the multi-part download logic. thanks

djdongjin commented on client/client.go:
```diff
@@ -394,9 +394,35 @@ type RemoteContext struct {
+	// - Downloading a layer
+	// - Unpacking a layer into the snapshotter
+	// - Moving an unpacked layer to its final location
+	MaxConcurrentDownloadOperations int
+
+	// MaxConcurrentDownloadsPerLayer controls parallel downloading of image layers
+	// by splitting them into chunks.
+	//
+	// - Values <= 1: Layers are downloaded using a single connection (default).
+	// - Values > 1: Layers are divided into chunks and downloaded in parallel,
+	//   which can significantly reduce pull times for large layers.
+	MaxConcurrentDownloadsPerLayer int
```
also on a side note: In the current PR we add this in 4 places: ctr, go client, cri, and transfer service.

Given it's a new feature and transfer service is used by default in containerd 2.0, I wonder if we should just add this only in transfer service. cc @dmcgowan

azr replied:
Adding this everywhere makes it easier to test things, using `ctr` & `cri`.

Also, I can see a scenario where someone wants to manually test the speed with their parameters depending on machine, etc.
And having it in `ctr`/`cri` kinda makes sense.
At which point I was like, yup adding it everywhere for consistency.

djdongjin replied:
For CRI specifically, I think the goal is to actually migrate existing config to transfer service, e.g.,

https://github.com/containerd/containerd/blob/d30521491d87d8172a4c922fff35e4afa660e63e/internal/cri/config/config.go#L311-L313

azr replied:
Oh, I see ! I think that to test and all, we should keep these, and the migration PR should move them.
But if then migration PR happens first, then I'll adapt the code. WDYT ?

djdongjin replied:
SGTM, fyi the migration PR is this one https://github.com/containerd/containerd/pull/8515

azr commented on client/client.go:
```diff
@@ -394,9 +394,35 @@ type RemoteContext struct {
+	// - Moving an unpacked layer to its final location
+	MaxConcurrentDownloadOperations int
+
+	// MaxConcurrentDownloadsPerLayer controls parallel downloading of image layers
+	// by splitting them into chunks.
+	//
+	// - Values <= 1: Layers are downloaded using a single connection (default).
+	// - Values > 1: Layers are divided into chunks and downloaded in parallel,
+	//   which can significantly reduce pull times for large layers.
+	MaxConcurrentDownloadsPerLayer int
```
MaxConcurrentDownloadsPerLayer is restricted by MaxConcurrentDownloads, yes.

So if you have `MaxConcurrentDownloads = 2`, the max value `MaxConcurrentDownloadsPerLayer` can effectively be is 2.

This part is define here:

https://github.com/containerd/containerd/blob/91289dfb1dfaaab76c62e552469d9b957f5046d3/core/remotes/resolver.go#L142-L153

djdongjin replied:
I feel making `MaxConcurrentDownloadsPerLayer` a per-layer config is easier to understand and make the code simpler, since `MaxConcurrentDownloads` and `MaxConcurrentDownloadsPerLayer` are independent and control the parallelism at different level.

Say if I only want muti-part per layer download, but not concurrent multi layer downloads, the current implementation cannot support this IIUC.

Or if previously I have `MaxConcurrentDownloads=3` and I want to enable multi part download, changing `MaxConcurrentDownloadsPerLayer` to, e.g., `3` won't change much (right?), as I need to figure out the new value for `MaxConcurrentDownloads` (probably 9).

azr commented on client/pull.go:
```diff
@@ -182,7 +182,23 @@ func (c *Client) fetch(ctx context.Context, rCtx *RemoteContext, ref string, lim
+		// here for a client pull, we limit the total number of concurrent
+		// downloads per pull.
+		opts = append(opts,
+			remotes.WithLimiter(semaphore.NewWeighted(int64(rCtx.MaxConcurrentDownloads))),
+			remotes.WithMaxConcurrentDownloads(rCtx.MaxConcurrentDownloads),
```
This is so that `open` can make a decision on the `MaxConcurrentDownloadsPerLayer` count ( `MaxConcurrentDownloadsPerLayer` needs to stay bellow `MaxConcurrentDownloads` )

djdongjin commented on client/client.go:
```diff
@@ -394,9 +394,35 @@ type RemoteContext struct {
+	// overall network bandwidth usage.
 	MaxConcurrentDownloads int
 
+	// MaxConcurrentDownloadOperations limits the number of parallel operations
+	// during an image pull to prevent system resource exhaustion.
+	//
+	// These operations include:
+	// - Downloading a layer
+	// - Unpacking a layer into the snapshotter
+	// - Moving an unpacked layer to its final location
+	MaxConcurrentDownloadOperations int
```
@azr can we actually remove this config? I feel this one is a bit confusing and not sure how to config an appropriate value. 

Previously we only have `MaxConcurrentDownloads` and IIUC we  didn't limit other ops like unpacking/moving? (actually can you explain difference between op Unpacking and op 3 Moving).

azr replied:
Trying, that's actually a think. I wanted to propose, maybe also the other configuration settings ( and just bool activate multipart_fetch ? ).
When deploying this containerd, it's a bit tedious to optimize a per node config config, and maybe we could infer that setting from cpu count, or something similar.

azr replied:
> actually can you explain difference between op Unpacking and op 3 Moving

When I was debugging things I saw this operation in the `handler` loop, now that I think about it, it shouldn't exist, should it ? removing this anyways, will try to debug and repro to explain better if time.

azr replied:
> actually can you explain difference between op Unpacking and op 3 Moving

There is no moving. If I got things correctly, we:
* open a bytestream to dl a file 
( sometimes that file is a manifest which will add more file to dl in step zero ) 
* put it in a tar
* open a bytestream to that tar
* apply a diff from it to a read-only fs

each individual step could be limited here.

azr replied:
Also re, about the preferred settings, when prodding this, I have been using the following setting everywhere:

```
    max_concurrent_downloads=32
    max_concurrent_downloads_per_layer=32
    concurrent_download_chunk_size=8388608
```

Instead of trying to guess a setting per machine. This setting proves to be bang on for almost every occasion, and we always end up being throttled by unpacking.
Some image downloads went from 2min to 25s/30s on big machines.

azr replied:
Please lmk if I can improve anything ! 😊

azr commented:
/retest

azr commented:
Ah, one test passed !

But another failed, not sure if related:

```
Captured StdOut/StdErr Output >>
  E0411 15:14:08.326569  110985 remote_runtime.go:193] "RunPodSandbox from runtime service failed" err=<
  	rpc error: code = Unknown desc = failed to setup network for sandbox "bd5aa2bbe77a7edb30e43ad0ec8301aff3c014bc0fff189d3c49140cb397419f": plugin type="portmap" failed (add): unable to create chain CNI-HOSTPORT-SETMARK: running [/usr/sbin/ip6tables -t nat -C CNI-HOSTPORT-SETMARK -m comment --comment CNI portfwd masquerade mark -j MARK --set-xmark 0x2000/0x2000 --wait]: exit status 2: ip6tables v1.8.7 (nf_tables): unknown option "--set-xmark"
  	Try `ip6tables -h' or 'ip6tables --help' for more information.
   >
  E0411 15:14:08.327505  110985 remote_runtime.go:222] "StopPodSandbox from runtime service failed" err="rpc error: code = Unknown desc = an error occurred when try to find sandbox \"\": prefix can't be empty" podSandboxID=""
  E0411 15:14:08.[327](https://github.com/containerd/containerd/actions/runs/14404946701/job/40402204401?pr=10177#step:19:328)893  110985 remote_runtime.go:241] "RemovePodSandbox from runtime service failed" err="rpc error: code = Unknown desc = an error occurred when try to find sandbox \"\": prefix can't be empty" podSandboxID=""
```
trying in case:

/retest

estesp commented:
> But another failed, not sure if related:

There is an ongoing issue with ipv6 networking and the current GH actions runner kernel; hopefully this will be solved soon.

dmcgowan commented on core/remotes/resolver.go:
```diff
@@ -43,7 +47,7 @@ type Resolver interface {
 	// All content fetched from the returned fetcher will be
 	// from the namespace referred to by ref.
-	Fetcher(ctx context.Context, ref string) (Fetcher, error)
+	Fetcher(ctx context.Context, ref string, opts ...FetcherOpt) (Fetcher, error)
```
I don't think we should make this interface change. I was trying to think of some better options. Previously we had another interface appended with `WithOptions`, but I don't think that should be necessary here. This configuration  could just be directly given as resolver options. The resolver could have a `SetOptions` function defined and type-asserted to in the transfer case. I think we can avoid changing the interface here though, especially since these settings are global and not defined per call.

azr replied:
Ack ! Updating !

azr replied:
Okay, updated, thanks for the review, I think this looks better, WDYT ?

azr commented:
/retest

azr commented:
Hey @dmcgowan & @djdongjin I updated the code to only add one setting: `ConcurrentLayerFetchBuffer`. When unset concurrent layer fetch will not happen. I removed the concurrent_layer fetch setting because in my test I ended up setting it to the same value as max concurrent downloads anyways and it was confusing.

Also I applied @dmcgowan's review to conf these settings ( `ConcurrentLayerFetchBuffer` and `MaxConcurrentDownloads` ) only once and not per download. Please let me know if that sounds like a better approach.

In the meantime, I just realised that testing 1 pull at a time was much faster that before. Put pulling two things at a time made for a longer total pull than if I were to pull at 2 different times, which doesn't sound like the best. Will investigate a bit.

azr commented:
re: 
> pulling two things at a time made for a longer total pull than if I were to pull at 2 different times

Turns out it's not the case on tmpfs, so I think this is okay.

So yeah okay this PR looks quite good to me, thanks for the many reviews, please let me know if anything needs to change.

Will try to have tests pass now.

dmcgowan commented on core/remotes/docker/fetcher.go:
```diff
@@ -262,74 +437,122 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest, op
+		}
+		go func() {
+			for i := range numChunks {
+				queue <- i
```
Should this also select on `stopChan` to prevent leaking of this go routine? The queue will stop draining once an error has occurred

azr replied:
Good point, adding !

azr replied:
done and tested a bunch, lmk if anything !

dmcgowan commented on core/remotes/docker/fetcher.go:
```diff
@@ -262,74 +437,122 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest, op
+							case <-stopChan:
+								return errors.New("another worker failed")
+							default:
+								close(stopChan)
```
This is still racy, multiple goroutines may hit this condition and cause a panic. There is no guarantee in one goroutines selects default and closes before another goroutine selects the default.

azr replied:
ditto here, I put the close in a sync.Once to be sure, also made the stop close the context. LMK if anything as well.

dmcgowan reviewed (changes requested):
I think these two conditions need to be checked to avoid leak and panic, but are hard to simulate in tests.

estesp commented:
/test pull-containerd-k8s-e2e-ec2

dmcgowan reviewed (approved):
We should just get this in, there are still maybe a few interface tweaks we can make before final release that won't affect the functionality.

fuweid commented on core/remotes/docker/fetcher.go:
```diff
@@ -262,74 +437,127 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest, op
+		r.Release(1)
+		for range parallelism {
+			go func() {
+				for i := range queue { // first in first out
```
Not sure if there is competive condition in http2 stdlib on window updating.
Will check this part later. it's not blocker.

azr commented on core/remotes/docker/resolver.go:
```diff
@@ -613,7 +646,75 @@ func (r *request) do(ctx context.Context) (*http.Response, error) {
+
+var errContentRangeIgnored = errors.New("content range requests ignored")
+
+var withOffsetCheck = func(offset int64) doChecks {
```
fixed !

azr commented on core/remotes/docker/fetcher.go:
```diff
@@ -262,74 +437,127 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest, op
-		if resp.StatusCode == http.StatusNotFound {
-			return nil, fmt.Errorf("content at %v not found: %w", req.String(), errdefs.ErrNotFound)
+	remaining, _ := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
+	if parallelism > 1 && chunkSize > 512 && req.body == nil {
```
good point, added, moved stuff around a bit

cartermckinnon commented:
@azr thank you for all the hard work on this, it's an awesome improvement! 🙏

azr commented:
Aw thanks ! ❤️ You're welcome !! 🚀
